Issue 915: Stop/start Location manager based on listeners registered.

soumyashisPR opened the pull request:
**Issue**: react-native-mapbox-gl https://github.com/react-native-mapbox-gl/maps/issues/915
This issue was caused by UserLocation component as it was stopping the locationManager, which can be shared by other UserLocation instances.

**Fix**:
'UserLocation' should only add/remove its listener on 'locationManager' service and never start or stop it. It the responsibility of the locationManager to start or stop itself depending on the number of listeners registered.
1. Removed code to start/stop locationManager from UserLocation component.
2. Added code to start/stop the LocationManager within LocationManager component itself, based on number of listeners registered at the moment.

------

EDIT ferdicus: 
resolves #915

soumyashisPR commented:
@mfazekas Would you be able to have a look at the PR? 
Thanks.

mfazekas commented on docs/UserLocation.md:
```diff
@@ -16,19 +16,19 @@
 | children | `any` | `none` | `false` | Custom location icon of type mapbox-gl-native components |
 
 ### methods
-#### setLocationManager({running})
```
This seems to be a breaking change, which means this could be merged into next major release. And we'll need a warning if they pass the old key `running`

soumyashisPR replied:
Hey @mfazekas , could you please guide me how we can handle this ?
I didn't follow how this would be a breaking change.

mfazekas replied:
My understanding is that before your change one called:
```setLocationManager({running: true})```

after your change:
```setLocationManager({required: true})```

So old code will break with the new libs, won't it?!

As for fixing, i see those options:

1. revert back to `running`
2. use `required` but when old code passes `running` use it as required value and show a warning.

soumyashisPR replied:
Thanks for the suggestions.
Reverted the rename.

mfazekas commented on javascript/components/UserLocation.js:
```diff
@@ -167,41 +165,39 @@ class UserLocation extends React.Component {
 
   async componentWillUnmount() {
     this._isMounted = false;
-    locationManager.removeListener(this._onLocationUpdate);
```
what happened with `removeListener` !?

soumyashisPR replied:
listener is removed/added in setLocationManager method, based on boolean 'required'.

mfazekas commented on javascript/modules/location/locationManager.js:
```diff
@@ -38,21 +38,24 @@ class LocationManager {
     if (!this._listeners.includes(listener)) {
       this._listeners.push(listener);
-
-      if (this._lastKnownLocation) {
```
why was this removed?!

soumyashisPR replied:
Change reverted. 
Reason for removing this, I was trying to keep addListener() and the method call listener() independent, but didn't work out.

soumyashisPR replied:
@mfazekas Please let me know if any more changes are required with this PR.
Thanks.

mfazekas reviewed (changes requested):
👍 Thanks, nice work.
It would be great if you could summarise the issues fixed with the changes. 

There are also some rename, which is not backward compatible. If the new names are better i'm all for it, but we must keep old versions working and showing a deprecation alert.

ferdicus commented on __tests__/components/UserLocation.test.js:
```diff
@@ -118,11 +118,12 @@ describe('UserLocation', () => {
-      fireEvent(shapeSource, 'onPress');
-      fireEvent(shapeSource, 'onPress');
-      expect(onPressCallback).toHaveBeenCalledTimes(2);
+      setTimeout(() => {
+        const shapeSource = UNSAFE_queryByType(ShapeSource);
+        fireEvent(shapeSource, 'onPress');
+        fireEvent(shapeSource, 'onPress');
+        expect(onPressCallback).toHaveBeenCalledTimes(2);
+      });
```
I know this got introduced in other tests, however it seemed to work without it before. 
What changed?

soumyashisPR replied:
Have reverted this change.
Had added this as fix to failing unit test.

ferdicus commented on __tests__/modules/location/locationManager.test.js:
```diff
@@ -88,17 +89,18 @@ describe('LocationManager', () => {
     });
 
     describe('#removeListener', () => {
+      MapboxGLLocationManager.stop = jest.fn();
+
```
why not test for it being called/ not called when you already added a jest mock?

soumyashisPR replied:
Added.

ferdicus commented on javascript/components/UserLocation.js:
```diff
@@ -167,41 +165,39 @@ class UserLocation extends React.Component {
    * @return {boolean}
    */
-  needsLocationManagerRunning() {
+  isLocationManagerRequired() {
```
maybe we could chose a different name here? 
It's awfully close to `_isLocationManagerRequired`

soumyashisPR replied:
Have reverted the rename changes.
Changed to `needsLocationManagerRunning`.

ferdicus reviewed (commented):
Thanks for tackling this, just added some comments.

Besides fixing failing tests it would be great if you would kindly add additional tests covering your changes. 

🙇

ferdicus commented:
@soumyashisPR, did you find the time to check the review comments?

Thanks 🙇

soumyashisPR commented:
Hey @ferdicus , have addressed the review comments. Let me know if there are further changes required. 
Currently stuck with some failing unit tests and unable to figure out the issue there.
Would be great if I could get some help / guidance.
Thanks

ferdicus commented on javascript/components/UserLocation.js:
```diff
@@ -167,41 +165,37 @@ class UserLocation extends React.Component {
-      if (running) {
-        locationManager.start();
-
-        const location = await locationManager.getLastKnownLocation();
```
I missed this the last time around, I think you can remove `getLastKnownLocation` then from `locationManager` as it's not being used anymore?

soumyashisPR replied:
Hi, this is actually being used in `setLocationManager` 
`if (running) {
        locationManager.addListener(this._onLocationUpdate);
        const location = await locationManager.getLastKnownLocation();
        this._onLocationUpdate(location);
      } else {
        locationManager.removeListener(this._onLocationUpdate);
      }`

ferdicus replied:
I mean... yeah, now it does after you reverted it again :)

ferdicus commented:
> Hey @ferdicus , have addressed the review comments. Let me know if there are further changes required.
> Currently stuck with some failing unit tests and unable to figure out the issue there.
> Would be great if I could get some help / guidance.
> Thanks

Thanks @soumyashisPR, can you please address all comments.  🙇 

Also, what seems to be the issue with the unit tests? 
Could you be more specific? 

Thanks again

systemlevel commented:
Is this PR going to create an issue for us who don't want the Location Manager running at all?  

For example, we are unmounting the <UserLocation /> When we don't want the location manager to run at all when the app is in the background, inactive, or active.

soumyashisPR commented:
> Is this PR going to create an issue for us who don't want the Location Manager running at all?
> 
> For example, we are unmounting the `<UserLocation />` when we don't want the location manager to run at all when the app is in the background, inactive, or active.

Hey @systemlevel, according to the PR, unmounting `<UserLocation />` will stop it from listening to the locationManager, it won't stop the location manager, unless it was the only `<UserLocation />`listening to the location manager, in which case it would stop.
Same applies for when `<UserLocation />` is mounted, it will start listening to locationManager, and location manager is started if not already started.

soumyashisPR commented:
> > Hey @ferdicus , have addressed the review comments. Let me know if there are further changes required.
> > Currently stuck with some failing unit tests and unable to figure out the issue there.
> > Would be great if I could get some help / guidance.
> > Thanks
> 
> Thanks @soumyashisPR, can you please address all comments. 🙇
> 
> Also, what seems to be the issue with the unit tests?
> Could you be more specific?
> 
> Thanks again

Nevermind, the issue with unit tests is fixed. 

Thanks.

mfazekas commented on __tests__/modules/location/locationManager.test.js:
```diff
@@ -80,7 +81,7 @@ describe('LocationManager', () => {
       });
 
-      test('does not readd same listener', () => {
+      test('does not read same listener', () => {
```
I think 're-add' make more sense here

soumyashisPR replied:
Fixed.

mfazekas commented on __tests__/modules/location/locationManager.test.js:
```diff
@@ -80,7 +81,7 @@ describe('LocationManager', () => {
       });
 
-      test('does not readd same listener', () => {
+      test('does not read same listener', () => {
```
```suggestion
      test('does not readd same listener', () => {
```

ferdicus reviewed (approved):
Thanks @soumyashisPR for this PR and sorry it took a while :)

Lets get this out the door 🚪
